From 3fd8f01814dc329be9334cc8a76d15be8c67d8d4 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Date: Wed, 8 Nov 2023 17:26:37 +0100
Subject: [PATCH 0620/1695] drivers: stm32_bsec3: add STM32MP21 version support

Support the BSEC IP version that is on STM32MP21x SoC
family.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git f63b9037bdf097ed1422aba97ec50e735d4b5f4f drivers: stm32_bsec3: add STM32MP21 version support]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I10c81761a281d35fe18c9eace7c9157633c713fb
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/361570
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/stm32_bsec3.c | 39 +++++++++++++++++++++++++++++++++-----
 1 file changed, 34 insertions(+), 5 deletions(-)

diff --git a/core/drivers/stm32_bsec3.c b/core/drivers/stm32_bsec3.c
index 3fd6e40ef..b019c50df 100644
--- a/core/drivers/stm32_bsec3.c
+++ b/core/drivers/stm32_bsec3.c
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: BSD-3-Clause
 /*
- * Copyright (c) 2022, STMicroelectronics
+ * Copyright (c) 2022-2024, STMicroelectronics
  */
 
 #include <assert.h>
@@ -84,10 +84,33 @@
 #define BSEC_DENR_SPIDENM		BIT(10)
 #define BSEC_DENR_SPNIDENM		BIT(11)
 #define BSEC_DENR_CFGSDIS		BIT(12)
+
+#if (defined(CFG_STM32MP25) || defined(CFG_STM32MP23))
 #define BSEC_DENR_CP15SDIS_MASK		GENMASK_32(14, 13)
 #define BSEC_DENR_CP15SDIS_SHIFT	13U
 #define BSEC_DENR_ALL_MASK		GENMASK_32(14, 1)
 #define BSEC_DENR_WRITE_CONF		U(0xDEB60000)
+#endif
+
+#if defined(CFG_STM32MP21)
+#define BSEC_DENR_ALL_MASK		GENMASK_32(17, 0)
+#define BSEC_DENR_WRITE_CONF		U(0xDEB00000)
+#endif
+
+// compute SECDED ECC as HAMMING(17,12) with parity
+#define PARITY_4BIT(x)		((((x) >> 3) ^ ((x) >> 2) ^ ((x) >> 1) ^ (x)) \
+				 & 1)
+#define PARITY_12BIT(x)		PARITY_4BIT(((x) >> 8) ^ ((x) >> 4) ^ (x))
+#define BSEC_DENR_ECC(x)	(((PARITY_12BIT((x) & 0x800)) << 17) | \
+				 ((PARITY_12BIT((x) & 0x7f0)) << 16) | \
+				 ((PARITY_12BIT((x) & 0x78e)) << 15) | \
+				 ((PARITY_12BIT((x) & 0x66d)) << 14) | \
+				 ((PARITY_12BIT((x) & 0xd5b)) << 13) | \
+				 ((PARITY_12BIT((x) ^ 0xcb7)) << 12))
+#define BSEC_DENR_v(x)		(BSEC_DENR_WRITE_CONF | BSEC_DENR_ECC(x) | \
+				 ((x) & 0xfff))
+#define BSEC_DENR_I		BSEC_DENR_v(0x0fff)
+#define BSEC_DENR_NI		BSEC_DENR_v(0x0fdf)
 
 /* BSEC_SR register fields */
 #define BSEC_SR_HVALID			BIT(1)
@@ -105,6 +128,7 @@
 #define BSEC_OTP_BANK_SHIFT		5U
 
 #define BSEC_IP_VERSION_1_0		U(0x10)
+#define BSEC_IP_VERSION_1_2		U(0x12)
 #define BSEC_IP_ID_3			U(0x100033)
 
 #define MAX_NB_TRIES			3U
@@ -512,11 +536,13 @@ TEE_Result stm32_bsec_write_debug_conf(uint32_t val)
 
 	exceptions = bsec_lock();
 
+	if (IS_ENABLED(CFG_STM32MP21))
+		val = BSEC_DENR_v(val);
+
 	io_clrsetbits32(bsec_base() + BSEC_DENR, BSEC_DENR_ALL_MASK,
 			val | BSEC_DENR_WRITE_CONF);
 
-
-	if (stm32_bsec_read_debug_conf() == val)
+	if (stm32_bsec_read_debug_conf() == (val & BSEC_DENR_ALL_MASK))
 		result = TEE_SUCCESS;
 
 	bsec_unlock(exceptions);
@@ -1153,8 +1179,11 @@ static TEE_Result initialize_bsec(void)
 
 	initialize_bsec_from_dt(fdt, node);
 
-	if ((bsec_get_version() != BSEC_IP_VERSION_1_0) ||
-	    (bsec_get_id() != BSEC_IP_ID_3))
+	if ((IS_ENABLED(CFG_STM32MP21) &&
+	     bsec_get_version() != BSEC_IP_VERSION_1_2) ||
+	    (IS_ENABLED(CFG_STM32MP25) &&
+	     bsec_get_version() != BSEC_IP_VERSION_1_0) ||
+	    bsec_get_id() != BSEC_IP_ID_3)
 		panic("BSEC probe wrong IP version/ID\n");
 
 	check_reset_error();
-- 
2.47.2

